File: Api/src/main/java/net/winternetwork/bedwars/api/module/Module.java
package net.winternetwork.bedwars.api.module;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

@Getter
@RequiredArgsConstructor
public abstract class Module {

    private final String name;
    private final ModulePriority priority;

    public void init() {
    }

    public void disable() {
    }

    public Runnable getScheduler() {
        return () -> {
        };
    }

    public void log(String text) {
        System.out.println(
                String.format("[%s] %s", getName(), text)
        );
    }

}

File: Api/src/main/kotlin/net/winternetwork/bedwars/api/module/Module.kt
package net.winternetwork.bedwars.api.module

import net.winternetwork.bedwars.api.module.manager.Manager
import java.io.File

abstract class Module(val name: String) {

    val dataFolder: File by lazy {
        File("plugins", name).apply {
            if (!exists()) mkdir()
        }
    }

    open val manager: Manager<*, *>? = null

    open fun init() {}
    open fun disable() {}
    open fun onSecPassed() {}

    fun log(text: String) {
        println(
                "[%s] %s".format(name, text)
        )
    }
}